docs/terrain.md

```diff
@@ -0,0 +1,123 @@
+
+# Contents
+
+- [Document Conventions](#document-conventions)
+- [File Structure](#file-structure)
+- [Data Types](#data-types)
+  - [Chunk](#chunk)
+  - [Voxel](#voxel)
+  - [Voxel.Flag](#voxel.flag)
+  - [Material](#material)
+
+## Document Conventions
+
+This document assumes a basic understanding of Rust's convention for numeric types. For example:
+
+- `u32` is an unsigned 32-bit integer
+- `f32` is a 32-bit floating point number
+
+All numeric types are little endian. Floats are stored as a dividend of their type's numeric maximum; as such, a `f8` with a value of `0x31` translates to `0x31 / 0xFF`, or `~0.192`. Floats are unsigned. Signed integers are stored with their absolute value, with preceding bytes containing the signedness information; meaning the range for an `i8` is `-0xFF` -> `0xFF`, instead of `-0x7F` to `0x7F`.
+
+Unless otherwise noted, all structs in this document are assumed to be stored with their components in the sequence listed without any modification. That is, if a struct is listed as being composed of an `i32` and an `f32`, it can be assumed that it is stored as an `i32` followed by an `f32`.
+
+## File Structure
+
+The first two bytes of the blob are `0x01`, which is a magic number, followed by `0x05`, which is a logarithm base 2 of the chunk size in voxels. Currently, the chunk size is 32<sup>3</sup> (32768). Values other than `0x05` are untested.
+
+Immediately following the header is an array of chunks, each of which must contain enough voxels to reach the maximum count (32<sup>3</sup>). Chunks are ascendingly ordered by X, then Y, then Z based on their position in the world. Each chunk represents a cube of 128<sup>3</sup> units in world space.
+
+| Field Name | Format          | Value                                                                 |
+| :--------- | :-------------- | :-------------------------------------------------------------------- |
+| Header     | `[u8; 2]`       | Magic number and chunk size.                                          |
+| Chunks     | [Chunk](#chunk) | As many chunks as needed to fill the blob. Unknown/no maximum amount. |
+
+## Data Types
+
+Terrain data is represented using a variety of different data types. Data described within them follows any conventions set in the [document conventions](#document-conventions).
+
+### Chunk
+
+The `Chunk` type is stored with a dynamic size dependent on the size of the voxels contained within, and the end of its data is marked by reaching its maximum voxel count of 32<sup>3</sup>. Voxels are ascendingly ordered by Y, then Z, then X based on their position in the chunk. Voxels are stored in rows of 32 units on each axis, each representing 4 units in world space.
```

Comment:
Is the maximum size fixed at 32<sup>3</sup> or is it dynamic based on the size specified in the header? It's unclear here because it says it's dynamically sized but the max value is set in stone.

Reply:
I'll test it real quick- I have no idea if the engine will overwrite the size in the header, if the game will crash, or what it will do.
32<sup>3</sup> is the default size, and can't be changed by normal means in Studio or any other developer tools

Reply:
Clarifying that changing from 32^3 / the value of 5 can not be loaded by the engine in new commit

Reply:
I lied it works above 5, when re-encoding/decoding it normalizes it back to 5/chunk size of 32 though it just splits up the big chunks